src/main/scala/com/fulcrumgenomics/bam/IdentifyPrimers.scala

```diff
@@ -0,0 +1,1027 @@
+
+  validate(maxMismatches >= 0,   "--max-mismatches must be >= 0")
+  validate(minMismatchDelta > 0, "--min-mismatch-delta must be > 0")
+  validate(matchScore > 0,       "--match-score must be > 0")
+  validate(mismatchScore < 0,    "--mismatch-score must be < 0")
+  validate(gapOpen < 0,          "--gap-open must be < 0")
+  validate(gapExtend < 0,        "--gap-extend must be < 0")
+  maxTemplatesInRam.foreach { m => validate(templatesPerThread < m, "--max-templates-in-ram must be greater than or equal to --templates-per-thread")}
+
+  private val aligner: Aligner = Aligner(matchScore = matchScore, mismatchScore = mismatchScore, gapOpen = gapOpen, gapExtend = gapExtend, mode = AlignmentMode.Glocal)
+
+  private val (fwdMatcher, revMatcher) = PrimerMatcher.build(
+    path              = this.primers,
+    aligner           = aligner,
+    slop              = slop,
+    maxMismatches     = maxMismatches,
+    minMismatchDelta  = minMismatchDelta,
+    minAlignmentScore = minAlignmentScore,
+    withFullAlignment = !skipFullAlignment,
+    revComp           = true
+  )
+
+  var numAlignments: AtomicLong = new AtomicLong(0)
+
+  private val PrimerPairMatchTypeTag: String = "pp"
+  private val PrimerInfoForwardTag: String   = "pf"
+  private val PrimerInfoReverseTag: String   = "pr"
+  private val NoPrimerMatchInfo: String      = "none"
+  private val comments: Seq[String]          = {
+    Seq(
+      "The pf/pr tags store the primer match metadata for the forward and reverse strand respectively.",
+      "The pf/pr tags are formatted as follow: <pair_id>,<primer_id>,<refName>:<start>-<end>,<strand>,<5-prime-offset>,<match-type>,<match-type-info>.",
+      s"The match-type is 'location', 'mismatch', or 'aligmnent' based on if the match was found using the location, mismatch-based alignment, or full-alignment.",
+      s"The pp tag is either 'canonical', 'non-canonical', 'single', or '$NoPrimerMatchInfo' based on if the matches for the primer pair agree, disagree, or are missing."
+    )
+  }
+
+  override def execute(): Unit = {
+    this.kswExecutable match {
+      case Some(p) => logger.info(s"Using ksw aligner: $p")
+      case None    => logger.info("Using the scala aligner.")
+    }
+    val metricCounter = new SimpleCounter[TemplateTypeMetric]()
+
+    // Group the reads by template
+    val in: SamSource = SamSource(this.input)
+    val iterator: Iterator[Template] = {
+      val iter = Bams.templateIterator(in)
+      numTemplates.map { n => iter.take(n) }.getOrElse { iter }
+    }
+
+    // NB: Add comments explaining tags in the output writer
+    val out: SamWriter = {
+      val header = {
+        val h = in.header.clone()
+        comments.foreach { comment => h.addComment(comment) }
+        h
+      }
+      SamWriter(path = output, header = header, sort = Some(SamOrder.Coordinate))
+    }
+
+    // Get the number of templates to process in a single thread
+    val batchSize = numTemplates.map(_ / threads) match {
```

Comment:
I don't think `numTemplates` should be used here.  I see why you're doing it, but it makes this whole section harder to follow, and it's essentially a debugging option in the first place, so if you're going to use that can we assume you're smart enough to set `templatesPerThread` and `threads` to sensible values?

I'd also ditch `maxTemplatesInRam` and just state that the program will keep up to `templatesPerThead *  readsPerTemplate * threads` reads in RAM.  Then you only need to calculate `majorBatchSize` and everything else is already defined.